File: java/jimmer-sql-graphql/service/src/main/java/org/babyfish/jimmer/sql/example/service/BookService.java
package org.babyfish.jimmer.sql.example.service;

import graphql.schema.DataFetchingEnvironment;
import org.babyfish.jimmer.spring.graphql.DataFetchingEnvironments;
import org.babyfish.jimmer.spring.model.SortUtils;
import org.babyfish.jimmer.spring.repository.SpringOrders;
import org.babyfish.jimmer.sql.example.model.Book;
import org.babyfish.jimmer.sql.example.repository.BookRepository;
import org.babyfish.jimmer.sql.example.service.dto.BookInput;
import org.babyfish.jimmer.sql.example.service.dto.BookSpecification;
import org.babyfish.jimmer.sql.example.service.dto.CompositeBookInput;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

/**
 * A real project should be a three-tier architecture consisting
 * of repository, service, and controller.
 *
 * This demo has no business logic, its purpose is only to tell users
 * how to use jimmer with the <b>least</b> code. Therefore, this demo
 * does not follow this convention, and let services be directly
 * decorated by `@Controller`, not `@Service`.
 */
@Controller
public class BookService {

    private final BookRepository bookRepository;

    public BookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    // --- Query ---

    @QueryMapping
    public List<Book> books(
            @Argument @Nullable String name,
            @Argument @Nullable BigDecimal minPrice,
            @Argument @Nullable BigDecimal maxPrice,
            @Argument @Nullable String storeName,
            @Argument @Nullable String authorName,
            @Argument @Nullable String sortCode,
            DataFetchingEnvironment env
    ) {
        Fetcher<Book> fetcher = DataFetchingEnvironments.createFetcher(Book.class, env);
        return bookRepository.findBooks(
                name,
                minPrice,
                maxPrice,
                storeName,
                authorName,
                sortCode,
                DataFetchingEnvironments.createFetcher(Book.class, env)
        );
    }

    @QueryMapping
    public List<Book> booksBySuperQBE(
            @Argument @Nullable BookSpecification specification,
            @Argument @Nullable String sortCode,
            DataFetchingEnvironment env
    ) {
        return bookRepository.find(
                specification,
                SortUtils.toSort(sortCode != null ? sortCode : "name asc"),
                DataFetchingEnvironments.createFetcher(Book.class, env)
        );
    }

    // --- Mutation ---

    @MutationMapping
    @Transactional
    public Book saveBook(@Argument BookInput input) {
        return bookRepository.save(input).getModifiedEntity();
    }

    @MutationMapping
    @Transactional
    public Book saveCompositeBook(@Argument CompositeBookInput input) {
        return bookRepository.save(input).getModifiedEntity();
    }

    @MutationMapping
    @Transactional
    public int deleteBook(@Argument long id) {
        bookRepository.deleteById(id);
        // GraphQL requires return value,
        // but `deleteById` of spring-data does not support value.
        // Is there a better design?
        return 1;
    }
}

File: kotlin/jimmer-sql-kt/service/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/service/BookService.kt
package org.babyfish.jimmer.sql.example.kt.service

import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.spring.model.SortUtils
import org.babyfish.jimmer.sql.example.kt.model.Book
import org.babyfish.jimmer.sql.example.kt.model.by
import org.babyfish.jimmer.sql.example.kt.repository.BookRepository
import org.babyfish.jimmer.sql.example.kt.service.dto.BookInput
import org.babyfish.jimmer.sql.example.kt.service.dto.BookSpecification
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.babyfish.jimmer.sql.exception.SaveException
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

/*
 * Why add spring web annotations to the service class?
 *
 * The success and popularity of rich client technologies represented by React, Vue and Angular
 * have greatly reduced the significance of the Controller layer on the spring server side.
 *
 * Moreover, over-bloated code structures are not conducive to demonstrating the capabilities
 * of the framework with small examples. Therefore, this example project no longer adheres to
 * dogmatism and directly adds spring web annotations to the service class.
 */
@RestController
@RequestMapping("/book")
@Transactional
class BookService(
    private val bookRepository: BookRepository
) {

    @GetMapping("/simpleList")
    fun findSimpleBooks(): List<@FetchBy("SIMPLE_FETCHER") Book> = // (1)
        bookRepository.findAll(SIMPLE_FETCHER) {
            asc(Book::name)
        }

    /**
     * The functionality of this method is the same as .[findBooksBySuperQBE]
     */
    @GetMapping("/list")
    fun findBooks(
        @RequestParam(defaultValue = "0") pageIndex: Int,
        @RequestParam(defaultValue = "5") pageSize: Int,
        // The `sortCode` also support implicit join, like `store.name asc`
        @RequestParam(defaultValue = "name asc, edition desc") sortCode: String,
        @RequestParam name: String?,
        @RequestParam minPrice: BigDecimal?,
        @RequestParam maxPrice: BigDecimal?,
        @RequestParam storeName: String?,
        @RequestParam authorName: String?,
    ): Page<@FetchBy("DEFAULT_FETCHER") Book> = // (2)
        bookRepository.findBooks(
            PageRequest.of(pageIndex, pageSize, SortUtils.toSort(sortCode)),
            name,
            minPrice,
            maxPrice,
            storeName,
            authorName,
            DEFAULT_FETCHER
        )

    /**
     * The functionality of this method is the same as .[findBooks]
     */
    @GetMapping("/listBySuperQBE")
    fun findBooksBySuperQBE(
        @RequestParam(defaultValue = "0") pageIndex: Int,
        @RequestParam(defaultValue = "5") pageSize: Int,
        // The `sortCode` also support implicit join, like `store.name asc`
        @RequestParam(defaultValue = "name asc, edition desc") sortCode: String,
        specification: BookSpecification
    ): Page<@FetchBy("DEFAULT_FETCHER") Book> =
        bookRepository.find(
            PageRequest.of(pageIndex, pageSize, SortUtils.toSort(sortCode)),
            specification,
            DEFAULT_FETCHER
        )

    @GetMapping("/{id}")
    fun findComplexBook(
        @PathVariable id: Long,
    ): @FetchBy("COMPLEX_FETCHER") Book? = // (3)
        bookRepository.findById(id, COMPLEX_FETCHER)

    @PutMapping
    @Throws(SaveException::class) // (4)
    fun saveBook(
        @RequestBody input: BookInput
    ): @FetchBy("DEFAULT_FETCHER") Book = // (5)
        bookRepository
            .saveCommand(input)
            .execute(DEFAULT_FETCHER)
            .modifiedEntity

    @DeleteMapping("/{id}")
    fun deleteBook(@PathVariable id: Long) {
        bookRepository.deleteById(id)
    }

    companion object {

        private val SIMPLE_FETCHER = newFetcher(Book::class).by {
            name()
        }

        private val DEFAULT_FETCHER = newFetcher(Book::class).by {

            allScalarFields()
            tenant(false)

            store {
                name()
            }
            authors {
                firstName()
                lastName()
            }
        }

        private val COMPLEX_FETCHER = newFetcher(Book::class).by {

            allScalarFields()
            tenant(false)

            store {
                allScalarFields()
                avgPrice()
            }
            authors {
                allScalarFields()
            }
        }
    }
}

/*----------------Documentation Links----------------
(1) (2) (3) https://babyfish-ct.github.io/jimmer-doc/docs/spring/client/api#declare-fetchby
(4) (6) https://babyfish-ct.github.io/jimmer-doc/docs/spring/client/error#allow-to-throw-all-exceptions-of-family
(5) (7) https://babyfish-ct.github.io/jimmer-doc/docs/mutation/save-command/input-dto/
---------------------------------------------------*/
